Failure after instrumentation of code which has stackmap frames and array of more than 7 dimensions

I am in the process of migrating a SW from Java 8 to Java 11. We are using JaCoCo for code coverage and I have found a problem when running the unit tests with JaCoCo tool (using ant). 

JaCoCo fails when it instantiates a class that contains a primitive array with more than 8 dimensions. Please do not ask about the motivation of having an 8 dimensional array .... legacy code I have to maintain.

I have reduced the code to the min that reproduces the problem. 

- The execution goes fine when I do not use JaCoCo/EclEmma or I use Java 8
- I have tested with two different JaCoCo versions , 0.8.2 and 0.8.3

### Steps to reproduce

**JaCoCo version**: tested with 0.8.2 and 0.8.3
**Operating system**: MacOS HighSierra 10.13.6 and CentOS Linux release 7.5.1804 
**Tool integration**: Ant 1.10.5 and Eclipse EclEmma 3.1.1.201809121651
**Java Versions**: OpenJDK 11.0.1 and OpenJDK 11.0.2

Use the following piece of code

```
import org.junit.Test;
public class PrimitiveArrayJacocoTest {

	/**
	 * Test with a boolean array
	 */
	@Test
	public void testBooleanArray() {
		// Just instantiate the class
		new BooleanArrayJacoco();		
	}
	
	/**
	 * Test with an integer array
	 */
	@Test
	public void testIntegerArray() {	
		// Just instantiate the class
		new IntegerArrayJacoco();	
	}
	
	public class BooleanArrayJacoco {
		
		public BooleanArrayJacoco() {
		    
			// Declare a 8 dimension boolean array , if the amount of dimensions is reduced test does not fail. 
		    boolean[][][][][][][][] array = new boolean[12][][][][][][][];
		    int idxDepth0 = 0; 
		    if(array[idxDepth0] == null) {
		    	array[0] = new boolean[4][][][][][][]; // Create new node with 4 array pointers
		    }
		}
		
	}
	
	public class IntegerArrayJacoco {
		
		public IntegerArrayJacoco() {
		    
			// Declare a 8 dimension integer array , if the amount of dimensions is reduced test does not fail. 
		    int[][][][][][][][] array = new int[12][][][][][][][];
		    int idxDepth0 = 0; 
		    if(array[idxDepth0] == null) {
		    	array[0] = new int[4][][][][][][]; // Create new node with 4 array pointers
		    }
		}
		
	}

}
```

### Expected behaviour

JaCoCo is able to run the tests with no failure.

### Actual behaviour

Tests fail with the following stack trace

```
java.lang.NoClassDefFoundError: I
 at gaia.cu1.tools.util.lookuptree.PrimitiveArrayJacocoTest.testIntegerArray(PrimitiveArrayJacocoTest.java:44)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:566)
 at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
 at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
 at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
 at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
 at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
 at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
 at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
 at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
 at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
 at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
 at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
 at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
 at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
Caused by: java.lang.ClassNotFoundException: I
 at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
 at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
 at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
 ... 24 more
```

and running the testBooleanArray .... the same but with different ClassNotFoundException:

```
java.lang.NoClassDefFoundError: Z
 at gaia.cu1.tools.util.lookuptree.PrimitiveArrayJacocoTest.testBooleanArray(PrimitiveArrayJacocoTest.java:33)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:566)
 at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
 at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
 at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
 at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
 at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
 at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
 at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
 at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
 at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
 at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
 at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
 at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
 at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
 at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
Caused by: java.lang.ClassNotFoundException: Z
 at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
 at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
 at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
 ... 24 more
```

Hi @quiquinSP, first of all thank you for the reproducer! :thumbsup: :heart:

However I doubt about correctness of your statement

> The execution goes fine when I use Java 8

because following reduced `Example.java`

```
public class Example {
    public static void main(String[] args) {
        boolean[][][][][][][][] a = new boolean[12][][][][][][][];
        int i = 0;
        if (a[i] == null) {
            a[0] = new boolean[4][][][][][][];
        }
        System.out.println("OK");
    }
}
```

demonstrates exactly the same problem even with Java 8:

```
$ java -version
java version "1.8.0_172"
Java(TM) SE Runtime Environment (build 1.8.0_172-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.172-b11, mixed mode)

$ javac -version
javac 1.8.0_172

$ javac -source 8 -target 8 Example.java

$  java -javaagent:jacoco-0.8.3/lib/jacocoagent.jar Example
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.NoClassDefFoundError: Z
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.privateGetMethodRecursive(Class.java:3048)
	at java.lang.Class.getMethod0(Class.java:3018)
	at java.lang.Class.getMethod(Class.java:1784)
	at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
Caused by: java.lang.ClassNotFoundException: Z
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 7 more
```

but the problem goes away when class has no stackmap frames, i.e. when compiled into Java 5 bytecode:

```
$ javap -v -p Example | grep "StackMapTable:" | wc -l
1

$ javap -v -p Example | grep "major"
  major version: 52

$ javac -source 5 -target 5 Example.java
warning: [options] bootstrap class path not set in conjunction with -source 1.5
warning: [options] source value 1.5 is obsolete and will be removed in a future release
warning: [options] target value 1.5 is obsolete and will be removed in a future release
warning: [options] To suppress warnings about obsolete options, use -Xlint:-options.
4 warnings

$ javap -v -p Example | grep "StackMapTable:" | wc -l
0

$ javap -v -p Example | grep "major"
  major version: 49

$ java -javaagent:jacoco-0.8.3/lib/jacocoagent.jar Example
OK
```

and seems to be related to the bug in ASM library - https://gitlab.ow2.org/asm/asm/issues/317866

As a temporary workaround you can exclude such classes from JaCoCo instrumentation.

--- a/org.jacoco.build/pom.xml
+++ b/org.jacoco.build/pom.xml
@@ -138,7 +138,7 @@
     <argLine>${jvm.args}</argLine>
 
     <!-- Dependencies versions -->
-    <asm.version>7.0</asm.version>
+    <asm.version>7.1</asm.version>
     <ant.version>1.7.1</ant.version>
     <args4j.version>2.0.28</args4j.version>
     <junit.version>4.8.2</junit.version>

--- a/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
+++ b/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
@@ -75,12 +75,12 @@ public class InstrSupportTest {
 		final byte[] bytes = new byte[] { //
 				(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic
 				0x00, 0x03, // minor_version
-				0x00, 0x2d, // major_version
+				(byte) 0x80, 0x12, // major_version
 				0x00, 0x02, // constant_pool_count
 				0x01, 0x00, 0x00 // constant_pool
 		};
 
-		assertEquals(45, InstrSupport.getVersionMajor(new ClassReader(bytes)));
+		assertEquals(32786, InstrSupport.getVersionMajor(new ClassReader(bytes)));
 	}
 
 	@Test

--- a/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
+++ b/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
@@ -95,7 +95,7 @@ public class InstrSupportTest {
 	}
 
 	@Test
-	public void getVersionMajor_should_return_major_version() {
+	public void getMajorVersion_should_read_major_version_from_ClassReader() {
 		final byte[] bytes = new byte[] { //
 				(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic
 				0x00, 0x03, // minor_version
@@ -104,7 +104,8 @@ public class InstrSupportTest {
 				0x01, 0x00, 0x00 // constant_pool
 		};
 
-		assertEquals(32786, InstrSupport.getVersionMajor(new ClassReader(bytes)));
+		assertEquals(32786,
+				InstrSupport.getMajorVersion(new ClassReader(bytes)));
 	}
 
 	@Test

--- a/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
+++ b/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
@@ -71,10 +71,16 @@ public class InstrSupportTest {
 	}
 
 	@Test
-	public void getVersionMajor_should_return_major_version_number() {
-		final byte[] bytes = new byte[] { (byte) 0xCA, (byte) 0xFE, (byte) 0xBA,
-				(byte) 0xBE, /* minor */ 0x00, 0x03, /* major */ 0x00, 0x2D };
-		assertEquals(45, InstrSupport.getVersionMajor(bytes));
+	public void getVersionMajor_should_return_major_version() {
+		final byte[] bytes = new byte[] { //
+				(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic
+				0x00, 0x03, // minor_version
+				0x00, 0x2d, // major_version
+				0x00, 0x02, // constant_pool_count
+				0x01, 0x00, 0x00 // constant_pool
+		};
+
+		assertEquals(45, InstrSupport.getVersionMajor(new ClassReader(bytes)));
 	}
 
 	@Test

--- a/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
+++ b/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
@@ -95,8 +95,9 @@ public class InstrSupportTest {
 	}
 
 	@Test
-	public void getMajorVersion_should_read_major_version_from_ClassReader() {
+	public void getMajorVersion_should_read_major_version_from_ClassReader_at_offset_relative_to_constant_pool() {
 		final byte[] bytes = new byte[] { //
+				(byte) 0xFF, // before class bytes
 				(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic
 				0x00, 0x03, // minor_version
 				(byte) 0x80, 0x12, // major_version
@@ -105,7 +106,7 @@ public class InstrSupportTest {
 		};
 
 		assertEquals(32786,
-				InstrSupport.getMajorVersion(new ClassReader(bytes)));
+				InstrSupport.getMajorVersion(new ClassReader(bytes, 1, -1)));
 	}
 
 	@Test

--- a/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
+++ b/org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java
@@ -71,6 +71,30 @@ public class InstrSupportTest {
 	}
 
 	@Test
+	public void getMajorVersion_should_read_unsigned_two_bytes_at_offset_6() {
+		final byte[] bytes = new byte[] { //
+				(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic
+				(byte) 0xFF, (byte) 0xFF, // minor_version
+				(byte) 0x80, (byte) 0x12 // major_version
+		};
+
+		assertEquals(32786, InstrSupport.getMajorVersion(bytes));
+	}
+
+	@Test
+	public void setMajorVersion_should_write_unsigned_two_bytes_at_offset_6() {
+		final byte[] bytes = new byte[8];
+
+		InstrSupport.setMajorVersion(32786, bytes);
+
+		assertArrayEquals(
+				new byte[] { (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, // magic
+						(byte) 0x00, (byte) 0x00, // minor_version
+						(byte) 0x80, (byte) 0x12 // major_version
+				}, bytes);
+	}
+
+	@Test
 	public void getVersionMajor_should_return_major_version() {
 		final byte[] bytes = new byte[] { //
 				(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE, // magic

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -81,7 +81,7 @@ public class Instrumenter {
 		};
 		final IProbeArrayStrategy strategy = ProbeArrayStrategyFactory
 				.createFor(classId, reader, accessorGenerator);
-		final int version = InstrSupport.getVersionMajor(reader);
+		final int version = InstrSupport.getMajorVersion(reader);
 		final ClassVisitor visitor = new ClassProbesAdapter(
 				new ClassInstrumenter(strategy, writer),
 				InstrSupport.needsFrames(version));

--- a/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
+++ b/org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java
@@ -81,7 +81,7 @@ public class Instrumenter {
 		};
 		final IProbeArrayStrategy strategy = ProbeArrayStrategyFactory
 				.createFor(classId, reader, accessorGenerator);
-		final int version = InstrSupport.getVersionMajor(source);
+		final int version = InstrSupport.getVersionMajor(reader);
 		final ClassVisitor visitor = new ClassProbesAdapter(
 				new ClassInstrumenter(strategy, writer),
 				InstrSupport.needsFrames(version));

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -169,7 +169,7 @@ public final class InstrSupport {
 	 * @see #setMajorVersion(int, byte[])
 	 * @see #getMajorVersion(ClassReader)
 	 */
-	static int getMajorVersion(final byte[] b) {
+	public static int getMajorVersion(final byte[] b) {
 		return ((b[6] & 0xFF) << 8) | (b[7] & 0xFF);
 	}
 
@@ -183,13 +183,13 @@ public final class InstrSupport {
 	 *            bytes of class
 	 * @see #getMajorVersion(byte[])
 	 */
-	static void setMajorVersion(final int majorVersion, final byte[] b) {
+	public static void setMajorVersion(final int majorVersion, final byte[] b) {
 		b[6] = (byte) (majorVersion >>> 8);
 		b[7] = (byte) majorVersion;
 	}
 
 	/**
-	 * Gets major of bytecode version number from given {@link ClassReader}.
+	 * Gets major version number from given {@link ClassReader}.
 	 *
 	 * @param reader
 	 *            reader to get information about the class

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -158,18 +158,16 @@ public final class InstrSupport {
 	 */
 	static final int CLINIT_ACC = Opcodes.ACC_SYNTHETIC | Opcodes.ACC_STATIC;
 
-	private static final int MAJOR_VERSION_INDEX = 6;
-
 	/**
-	 * Gets major of bytecode version number from given bytes of class.
+	 * Gets major of bytecode version number from given {@link ClassReader}.
 	 *
-	 * @param b
-	 *            bytes of class
-	 * @return version of bytecode
+	 * @param reader
+	 *            reader to get information about the class
+	 * @return major version of bytecode
 	 */
-	public static int getVersionMajor(final byte[] b) {
-		return (short) (((b[MAJOR_VERSION_INDEX] & 0xFF) << 8)
-				| (b[MAJOR_VERSION_INDEX + 1] & 0xFF));
+	public static int getVersionMajor(final ClassReader reader) {
+		final int firstConstantPoolEntryOffset = reader.getItem(1) - 1;
+		return reader.readShort(firstConstantPoolEntryOffset - 4);
 	}
 
 	/**
@@ -238,7 +236,7 @@ public final class InstrSupport {
 	 */
 	public static ClassReader classReaderFor(final byte[] b) {
 		final byte[] originalVersion = new byte[] { b[4], b[5], b[6], b[7] };
-		if (getVersionMajor(b) == Opcodes.V12 + 1) {
+		if (b[6] == 0x00 && b[7] == 0x39) {
 			b[4] = (byte) (Opcodes.V12 >>> 24);
 			b[5] = (byte) (Opcodes.V12 >>> 16);
 			b[6] = (byte) (Opcodes.V12 >>> 8);

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -194,8 +194,12 @@ public final class InstrSupport {
 	 * @param reader
 	 *            reader to get information about the class
 	 * @return major version of bytecode
+	 * @see ClassReader#ClassReader(byte[], int, int)
+	 * @see #getMajorVersion(byte[])
 	 */
 	public static int getMajorVersion(final ClassReader reader) {
+		// relative to the beginning of constant pool because ASM provides API
+		// to construct ClassReader which reads from the middle of array
 		final int firstConstantPoolEntryOffset = reader.getItem(1) - 1;
 		return reader.readUnsignedShort(firstConstantPoolEntryOffset - 4);
 	}
@@ -267,6 +271,7 @@ public final class InstrSupport {
 	public static ClassReader classReaderFor(final byte[] b) {
 		final int originalVersion = getMajorVersion(b);
 		if (originalVersion == Opcodes.V12 + 1) {
+			// temporarily downgrade version to bypass check in ASM
 			setMajorVersion(Opcodes.V12, b);
 		}
 		final ClassReader classReader = new ClassReader(b);

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -167,7 +167,7 @@ public final class InstrSupport {
 	 */
 	public static int getVersionMajor(final ClassReader reader) {
 		final int firstConstantPoolEntryOffset = reader.getItem(1) - 1;
-		return reader.readShort(firstConstantPoolEntryOffset - 4);
+		return reader.readUnsignedShort(firstConstantPoolEntryOffset - 4);
 	}
 
 	/**

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -159,6 +159,36 @@ public final class InstrSupport {
 	static final int CLINIT_ACC = Opcodes.ACC_SYNTHETIC | Opcodes.ACC_STATIC;
 
 	/**
+	 * Gets major version number from given bytes of class (unsigned two bytes
+	 * at offset 6).
+	 *
+	 * @return major version of bytecode
+	 * @see <a href=
+	 *      "https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1">Java
+	 *      Virtual Machine Specification §4 The class File Format</a>
+	 * @see #setMajorVersion(int, byte[])
+	 * @see #getVersionMajor(ClassReader)
+	 */
+	static int getMajorVersion(final byte[] b) {
+		return ((b[6] & 0xFF) << 8) | (b[7] & 0xFF);
+	}
+
+	/**
+	 * Sets major version number in given bytes of class (unsigned two bytes at
+	 * offset 6).
+	 *
+	 * @param majorVersion
+	 *            major version of bytecode to set
+	 * @param b
+	 *            bytes of class
+	 * @see #getMajorVersion(byte[])
+	 */
+	static void setMajorVersion(final int majorVersion, final byte[] b) {
+		b[6] = (byte) (majorVersion >>> 8);
+		b[7] = (byte) majorVersion;
+	}
+
+	/**
 	 * Gets major of bytecode version number from given {@link ClassReader}.
 	 *
 	 * @param reader
@@ -235,15 +265,12 @@ public final class InstrSupport {
 	 * @return {@link ClassReader}
 	 */
 	public static ClassReader classReaderFor(final byte[] b) {
-		final byte[] originalVersion = new byte[] { b[4], b[5], b[6], b[7] };
-		if (b[6] == 0x00 && b[7] == 0x39) {
-			b[4] = (byte) (Opcodes.V12 >>> 24);
-			b[5] = (byte) (Opcodes.V12 >>> 16);
-			b[6] = (byte) (Opcodes.V12 >>> 8);
-			b[7] = (byte) Opcodes.V12;
+		final int originalVersion = getMajorVersion(b);
+		if (originalVersion == Opcodes.V12 + 1) {
+			setMajorVersion(Opcodes.V12, b);
 		}
 		final ClassReader classReader = new ClassReader(b);
-		System.arraycopy(originalVersion, 0, b, 4, originalVersion.length);
+		setMajorVersion(originalVersion, b);
 		return classReader;
 	}
 

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -162,6 +162,8 @@ public final class InstrSupport {
 	 * Gets major version number from given bytes of class (unsigned two bytes
 	 * at offset 6).
 	 *
+	 * @param b
+	 *            bytes of class
 	 * @return major version of bytecode
 	 * @see <a href=
 	 *      "https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1">Java

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/InstrSupport.java
@@ -167,7 +167,7 @@ public final class InstrSupport {
 	 *      "https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.1">Java
 	 *      Virtual Machine Specification §4 The class File Format</a>
 	 * @see #setMajorVersion(int, byte[])
-	 * @see #getVersionMajor(ClassReader)
+	 * @see #getMajorVersion(ClassReader)
 	 */
 	static int getMajorVersion(final byte[] b) {
 		return ((b[6] & 0xFF) << 8) | (b[7] & 0xFF);
@@ -195,7 +195,7 @@ public final class InstrSupport {
 	 *            reader to get information about the class
 	 * @return major version of bytecode
 	 */
-	public static int getVersionMajor(final ClassReader reader) {
+	public static int getMajorVersion(final ClassReader reader) {
 		final int firstConstantPoolEntryOffset = reader.getItem(1) - 1;
 		return reader.readUnsignedShort(firstConstantPoolEntryOffset - 4);
 	}

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/ProbeArrayStrategyFactory.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/ProbeArrayStrategyFactory.java
@@ -43,7 +43,7 @@ public final class ProbeArrayStrategyFactory {
 			final IExecutionDataAccessorGenerator accessorGenerator) {
 
 		final String className = reader.getClassName();
-		final int version = InstrSupport.getVersionMajor(reader.b);
+		final int version = InstrSupport.getVersionMajor(reader);
 
 		if (isInterfaceOrModule(reader)) {
 			final ProbeCounter counter = getProbeCounter(reader);

--- a/org.jacoco.core/src/org/jacoco/core/internal/instr/ProbeArrayStrategyFactory.java
+++ b/org.jacoco.core/src/org/jacoco/core/internal/instr/ProbeArrayStrategyFactory.java
@@ -43,7 +43,7 @@ public final class ProbeArrayStrategyFactory {
 			final IExecutionDataAccessorGenerator accessorGenerator) {
 
 		final String className = reader.getClassName();
-		final int version = InstrSupport.getVersionMajor(reader);
+		final int version = InstrSupport.getMajorVersion(reader);
 
 		if (isInterfaceOrModule(reader)) {
 			final ProbeCounter counter = getProbeCounter(reader);

--- a/org.jacoco.doc/docroot/doc/changes.html
+++ b/org.jacoco.doc/docroot/doc/changes.html
@@ -30,12 +30,25 @@
       (GitHub <a href="https://github.com/jacoco/jacoco/issues/849">#849</a>).</li>
 </ul>
 
+<h3>Fixed bugs</h3>
+<ul>
+  <li>Fixed incorrect update of frames caused by bug in ASM library in case of
+      arrays with more than 7 dimensions
+      (GitHub <a href="https://github.com/jacoco/jacoco/issues/839">#839</a>).</li>
+</ul>
+
 <h3>API Changes</h3>
 <ul>
     <li>Methods <code>Instrumenter.instrument(org.objectweb.asm.ClassReader)</code>
         and <code>Analyzer.analyzeClass(org.objectweb.asm.ClassReader)</code>
         were removed
         (GitHub <a href="https://github.com/jacoco/jacoco/issues/850">#850</a>).</li>
+</ul>
+
+<h3>Non-functional Changes</h3>
+<ul>
+  <li>JaCoCo now depends on ASM 7.1
+      (GitHub <a href="https://github.com/jacoco/jacoco/issues/851">#851</a>).</li>
 </ul>
 
 <h2>Release 0.8.3 (2019/01/23)</h2>

